src/m2n/PointToPointCommunication.cpp

```diff
@@ -236,6 +236,42 @@ void printLocalIndexCountStats(std::map<int, std::vector<int>> const &m)
   }
 }
 
+namespace {
+/**
+   * @brief This function is by by and large the same as std::set_intersection().
+   * The only difference is that we don't return the intersection set itself, but
+   * we return the indices of elements in \p InputIt1, which appear in both sets
+   * ( \p InputIt1 and \p InputIt2 )
+   * The implementation was taken from
+   * https://en.cppreference.com/w/cpp/algorithm/set_intersection#Version_1 with the
+   * only difference that we compute and store std::distance() in the output iterator.
+   * Similar to the std function, this function operates on sorted ranges.
+   *
+   * @param ref1 The reference iterator, to which we compute the distance/indices.
+   * @param first1 The begin of the first range we want to compute the intersection with
+   * @param last1 The end of the first range we want to compute the intersection with
+   * @param first1 The begin of the second range we want to compute the intersection with
+   * @param last1 The end of the second range we want to compute the intersection with
+   * @param d_first Beginning of the output range
+   * @return OutputIt Iterator past the end of the constructed range.
+   */
+template <class InputIt1, class InputIt2, class OutputIt>
+OutputIt set_intersection_indices(InputIt1 ref1, InputIt1 first1, InputIt1 last1,
+                                  InputIt2 first2, InputIt2 last2, OutputIt d_first)
+{
+  while (first1 != last1 && first2 != last2) {
+    if (*first1 < *first2)
+      ++first1;
```

Comment:
Not sure if this helps, but it could be interesting to see if changing this "step first1 once" to "advance first1 until" results in a performance uplift. Definitely low prio though. 

```suggestion
    if (*first1 < *first2) {
      while(first1 != last1 && *first1 < *first2) {
        ++first1;
      }
    }
```

Reply:
I would expect that the final behavior depends on the structure of the underlying data, so let's don't investigate this (unless you want to), see also comment below.